src/jit/morph.cpp

```diff
@@ -4766,15 +4766,33 @@ GenTreePtr Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPtr f
 #endif
 
 #ifdef _TARGET_ARM_
-    if (fgEntryPtr->isSplit)
+    if ((fgEntryPtr->isSplit && fgEntryPtr->numSlots + fgEntryPtr->numRegs > 4) ||
+        (!fgEntryPtr->isSplit && fgEntryPtr->regNum == REG_STK))
     {
-        if (fgEntryPtr->numSlots + fgEntryPtr->numRegs > 4)
+        // If already OBJ it is set properly already.
+        if (arg->OperGet() == GT_OBJ)
         {
             return arg;
         }
-    }
-    else if (!fgEntryPtr->isHfaRegArg && fgEntryPtr->numSlots > 4)
-    {
+
+        GenTreeLclVarCommon* lclCommon =
+            arg->OperGet() == GT_ADDR ? arg->gtOp.gtOp1->AsLclVarCommon() : arg->AsLclVarCommon();
+
+        // If we didn't change the type of the struct, it means
+        // its classification doesn't support to be passed directly through a
+        // register, so we need to pass a pointer to the destination where
+        // where we copied the struct to.
+        // Make sure this is an addr node.
+        if (arg->OperGet() != GT_ADDR && arg->OperGet() != GT_LCL_VAR_ADDR)
```

Comment:
The above comment doesn't seem to be relevant here. I think it would be better simply to say:
```
// We need to construct a `GT_OBJ` node for the argument, so we need to get the address of the lclVar.
```